File: routers/sub_classification.go
package routers

import (
	"book-keeping-web/models"
	"book-keeping-web/utils"

	"github.com/gin-gonic/gin"
)

type SubClassificationReq struct {
	Name             string `json:"name" binding:"omitempty"`
	ClassificationId int    `json:"classification_id"`
}

// SubClassificationCreate 为当前用户创建分类
func SubClassificationCreate(c *gin.Context) {
	req := SubClassificationReq{}
	if err := c.BindJSON(&req); err != nil {
		utils.FailMessage(c, "parse param error")
		return
	}
	classification := models.SubClassification{Name: req.Name, UserId: c.GetInt("UserId"), ClassificationId: req.ClassificationId}
	if !classification.CreateSubClassification() {
		utils.FailMessage(c, "create classification error")
		return
	}
	utils.SuccessMessage(c, "create classification success")
}

// SubClassificationList 查询当前用户的所有分类
func SubClassificationList(c *gin.Context) {
	req := SubClassificationReq{}
	if err := c.BindJSON(&req); err != nil {
		utils.FailMessage(c, "parse param error")
		return
	}
	var classifications []models.SubClassification
	classification := models.SubClassification{UserId: c.GetInt("UserId"), ClassificationId: req.ClassificationId}
	if !classification.ListSubClassification(&classifications) {
		utils.FailMessage(c, "list classifications error")
		return
	}
	res := classification.ToSubClassificationView(&classifications)
	utils.Success(c, gin.H{"sub_classifications": res}, "list classifications success")
}

// SubClassificationDelete 删除分类
func SubClassificationDelete(c *gin.Context) {
	req := SubClassificationReq{}
	if err := c.BindJSON(&req); err != nil {
		utils.FailMessage(c, "parse param error")
		return
	}
	classification := models.SubClassification{Name: req.Name, UserId: c.GetInt("UserId"), ClassificationId: req.ClassificationId}
	if !classification.DeleteSubClassification() {
		utils.FailMessage(c, "delete classification error")
		return
	}
	utils.SuccessMessage(c, "delete classification success")
}

File: frontend/src/api/sub_classification.ts
import axios from './instance/bff_instance'
import { ClassificationId, SubClassificationItem } from "./interface";

const subClassificationApi = {
  listSubClassification,
  createSubClassification,
  deleteSubClassification,
}

/**
 * 查询分类列表
 */
export async function listSubClassification(classification_id: ClassificationId): Promise<SubClassificationItem[]> {
  try {
    const response = await axios.post<{ sub_classifications: SubClassificationItem[] }>('/sub_classification/list', {
      classification_id: classification_id, // 分类ID
    });
    return response.data.sub_classifications;
  } catch (error) {
    console.error('listSubClassification error', error)
    return [];
  }
}


/**
 * 创建分类
 */
export async function createSubClassification(classification_id: ClassificationId, name: string): Promise<any> {
  try {
    const response = await axios.post<any>('/sub_classification/create', {
      classification_id: classification_id, // 分类ID
      name: name, // 子分类名称
    });
    return response.data;
  } catch (error) {
    console.error('createSubClassification error', error)
    return null;
  }
}


/**
 * 删除分类
 */
export async function deleteSubClassification(classification_id: ClassificationId, name: string): Promise<boolean> {
  try {
    await axios.post<any>('/sub_classification/delete', {
      classification_id: classification_id, // 分类ID
      name: name, // 子分类名称
    });
    return true;
  } catch (error) {
    console.error('deleteLedger error', error)
    return false;
  }
}

export default subClassificationApi